### PR TITLE
Expose flag for critical extensions

### PR DESCRIPTION
**Pull request by creiter32**
## Description

Changes the API to enable creating X.509 CSRs with critical extensions. Fixes https://github.com/ARMmbed/mbedtls/issues/4055

## Status
**READY**

## Requires Backporting
NO

## Additional comments

I didn't find any test cases for `mbedtls_x509write_csr_set_extension` so I'm not sure if this is needed.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Thank you for your contribution! This feature makes sense to me, and the code is ok.

This is an API change: any code currently using `mbedtls_x509write_csr_set_extension` will no longer compile. We don't do API changes in Mbed TLS except when there's a major version change. Fortunately, we're getting ready to go from Mbed TLS 2.x to Mbed TLS 3.0, so if this can be merged in the next couple of weeks, the API change is ok.

A changelog entry and a migration guide entry are required. See [`ChangeLog.d/00README.md`](https://github.com/ARMmbed/mbedtls/blob/development/ChangeLog.d/00README.md) and [`docs/3.0-migration-guide.md`](https://github.com/ARMmbed/mbedtls/blob/development/docs/3.0-migration-guide.md).

It would be better to have tests, but given the preexisting lack of tests for `mbedtls_x509write_csr_set_extension` I won't insist on it. A contribution of unit tests for CSR creation with both critical and non-critical extensions would be appreciated though.

**Comment by creiter32** [comment]
@gilles-peskine-arm I have added the changelog and migration guide parts.

File: docs/3.0-migration-guide.d/csr-add-critical-extension.md
Comment (gilles-peskine-arm) [inline docs/3.0-migration-guide.d/csr-add-critical-extension.md:4]:
```suggestion
This affects applications that call the `mbedtls_x509write_csr_set_extension`
```

File: ChangeLog.d/issue4055.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/issue4055.txt:3]:
```suggestion
   * The function mbedtls_x509write_csr_set_extension() has an extra parameter
     which allows marking an extension as critical. Fixes #4055.
```

File: include/mbedtls/x509_csr.h
Comment (gilles-peskine-arm) [inline include/mbedtls/x509_csr.h:238]:
```suggestion
 * \param critical  Set to 1 to mark the extension as critical, 0 otherwise.
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks good to me except some minor English improvements in the documentation.

I've filed the lack of tests in https://github.com/ARMmbed/mbedtls/issues/4577. Since it's preexisting, it doesn't block this pull request.

**Review by gilles-peskine-arm — APPROVED** [review]
Thanks for the changes, looks good to me!

The next step in the process is a review by another maintainer. If the second reviewer requests changes, please make them in separate commits, not by amending the current commit. For future reference, making successive commits is the general rule: amending commits should be exceptional.

**Review by d3zd3z — APPROVED** [review]
I agree with @gilles-peskine-arm review, the code looks fine, testing would be nice, but since we don't have any surrounding testing is probably too much to ask for here, and yes, let's try to get the API change in while that can be done.

**Pull request merged**
